### DISCUSSION TITLE
Patch pcode in script

### DISCUSSION THREAD

**Discussion by aiglematth** [NationalSecurityAgency/ghidra#5693] [Q&A]
Hello,
I am working on a Ghidra script to deobfuscate certain programs. As part of this, I want to replace a particular function call with multiple jump instructions. I was considering using pcode patching because it's architecture-agnostic and avoids issues that can arise with direct assembly patching. However, I'm struggling to modify the pcode displayed in the code browser. Does anyone have suggestions or a code snippet that demonstrates how to edit the pcode shown in the code browser? Thank you!

**Comment by sad-dev** [comment]
https://github.com/StarCrossPortal/ghidracraft is dead but pretty much did what you are asking for

**Comment by aiglematth** [comment]
> https://github.com/StarCrossPortal/ghidracraft is dead but pretty much did what you are asking for

Thanks for your answer.
Yeah I saw it, but is there any way to do it from the main version of ghidra ?

**Comment by astrelsky** [comment]
If you're replacing a function call, use a call fixup.

**Comment by aiglematth** [comment]
> If you're replacing a function call, use a call fixup.

Thx.
And if I am replacing load instructions ?

**Comment by astrelsky** [comment]
> > If you're replacing a function call, use a call fixup.
> 
> Thx.
> And if I am replacing load instructions ?

To my knowledge you would have no other choice than to patch the instruction.

**Comment by aiglematth** [comment]
> > > If you're replacing a function call, use a call fixup.
> > 
> > Thx.
> > And if I am replacing load instructions ?
> 
> To my knowledge you would have no other choice than to patch the instruction.

Ok thanks. So no way to transform a call pcode in cbranch pcode in your opinion?

**Comment by astrelsky** [comment]
> > > > If you're replacing a function call, use a call fixup.
> > > 
> > > Thx.
> > > And if I am replacing load instructions ?
> > 
> > To my knowledge you would have no other choice than to patch the instruction.
> 
> Ok thanks. So no way to transform a call pcode in cbranch pcode in your opinion?

You can use a call fixup to replace every call to a specific function with whatever pcode you want.

https://github.com/NationalSecurityAgency/ghidra/blob/master/Ghidra/Processors/x86/data/languages/x86-64-gcc.cspec#L365

You can add these per program by going to edit->Options for _ProgramName_-> specification extensions

**Comment by aiglematth** [comment]
> > > > > If you're replacing a function call, use a call fixup.
> > > > 
> > > > Thx.
> > > > And if I am replacing load instructions ?
> > > 
> > > To my knowledge you would have no other choice than to patch the instruction.
> > 
> > Ok thanks. So no way to transform a call pcode in cbranch pcode in your opinion?
> 
> You can use a call fixup to replace every call to a specific function with whatever pcode you want.
> 
> https://github.com/NationalSecurityAgency/ghidra/blob/master/Ghidra/Processors/x86/data/languages/x86-64-gcc.cspec#L365
> 
> You can add these per program by going to edit->Options for _ProgramName_-> specification extensions

Ohhh ! This can help me. Do you know if it is possible to set the pcode dynamic="true" attribute and if yes how to inject the pcode ? Otherwise, is it possible to set the pcodeinjectlibrary without modify the processor spec file (from a script for example) ?

**Reply by astrelsky** [reply]
You don't need to do that. I didn't have my laptop on hand this morning so I wasn't able to give a real example.

I have attached an actual call fixup that I have laying around (which implies I've used it at least once). The `.txt` extension is just to circumvent GitHub's unnecessary file extension restriction.

[__aeabi_uldivmod.xml.txt](https://github.com/NationalSecurityAgency/ghidra/files/12403000/__aeabi_uldivmod.xml.txt)
